# Preprocess Discharge Summaries

In [2]:
import pandas as pd
import numpy as np
import os
import psycopg2
import sqlalchemy
import string
import spacy
from spacy.symbols import ORTH
import scispacy
from collections import Counter
import re
from datetime import date, datetime, timedelta
import random
from sklearn.model_selection import GroupShuffleSplit, StratifiedShuffleSplit
from spellchecker import SpellChecker
from flashtext import KeywordProcessor
import nltk

Connect to the mimic database and set the search path to the 'mimiciii' schema

In [3]:
dbschema='mimiciii'
cnx = sqlalchemy.create_engine('postgresql+psycopg2://aa5118:mimic@localhost:5432/mimic',
                    connect_args={'options': '-csearch_path={}'.format(dbschema)})


Query the discharge summary notes joined on to patient data

In [4]:
sql = """
  SELECT
      p.subject_id, p.dob, p.gender,
      n.hadm_id, n.category, n.chartdate, n.row_id,
      ROUND((cast(chartdate as date) - cast(dob as date)) / 365.242,0)
          AS age_at_noteevent,
      n.text
  FROM patients p 
  INNER JOIN noteevents n 
  ON p.subject_id = n.subject_id
  WHERE ROUND((cast(chartdate as date) - cast(dob as date)) / 365.242,0) > 14
  AND n.category = 'Discharge summary'
  ORDER BY subject_id, chartdate DESC
  --LIMIT 25000
"""

df = pd.read_sql_query(sqlalchemy.text(sql), cnx)
df.head()

subject_id        dob gender  hadm_id           category  chartdate  \
0           3 2025-04-11      M   145834  Discharge summary 2101-10-31   
1           4 2143-05-12      F   185777  Discharge summary 2191-03-23   
2           6 2109-06-21      F   107064  Discharge summary 2175-06-15   
3           9 2108-01-26      M   150750  Discharge summary 2149-11-14   
4           9 2108-01-26      M   150750  Discharge summary 2149-11-13   

   row_id  age_at_noteevent                                               text  
0   44005              77.0  Admission Date:  [**2101-10-20**]     Discharg...  
1    4788              48.0  Admission Date:  [**2191-3-16**]     Discharge...  
2   20825              66.0  Admission Date: [**2175-5-30**]        Dischar...  
3   57115              42.0  Name:  [**Known lastname 10050**], [**Known fi...  
4   20070              42.0  Admission Date:  [**2149-11-9**]       Dischar...

Change data type of age to the smallest possible type of integer to save memory and get rid of decimal point

In [4]:
df['age_at_noteevent'] = pd.to_numeric(df['age_at_noteevent'], downcast='integer')
df.head(20)

subject_id        dob gender  hadm_id           category  chartdate  \
0            3 2025-04-11      M   145834  Discharge summary 2101-10-31   
1            4 2143-05-12      F   185777  Discharge summary 2191-03-23   
2            6 2109-06-21      F   107064  Discharge summary 2175-06-15   
3            9 2108-01-26      M   150750  Discharge summary 2149-11-14   
4            9 2108-01-26      M   150750  Discharge summary 2149-11-13   
5           11 2128-02-22      F   194540  Discharge summary 2178-05-11   
6           12 2032-03-24      M   112213  Discharge summary 2104-08-20   
7           13 2127-02-27      F   143045  Discharge summary 2167-01-15   
8           13 2127-02-27      F   143045  Discharge summary 2167-01-15   
9           17 2087-07-14      F   161087  Discharge summary 2135-05-13   
10          17 2087-07-14      F   194023  Discharge summary 2134-12-31   
11          18 2116-11-29      M   188822  Discharge summary 2167-10-04   
12          18 2116-11-29      M   188822  Discharge summary 2167-10-02   
13          18 2116-11-29      M   188822  Discharge summary 2167-10-02   
14          19 1808-08-05      M   109235  Discharge summary 2108-08-11   
15          20 2107-06-13      F   157681  Discharge summary 2183-05-03   
16          20 2107-06-13      F   157681  Discharge summary 2183-04-28   
17          21 2047-04-04      M   111970  Discharge summary 2135-02-08   
18          21 2047-04-04      M   109451  Discharge summary 2134-09-24   
19          22 2131-05-07      F   165315  Discharge summary 2196-04-10   

    row_id  age_at_noteevent  \
0    44005                77   
1     4788                48   
2    20825                66   
3    57115                42   
4    20070                42   
5    30120                50   
6    50972                72   
7    57099                40   
8    20168                40   
9    51783                48   
10   51782                47   
11    7824                51   
12    7823                51   
13   55875                51   
14   25231               300   
15   17861                76   
16   17860                76   
17    7238                88   
18    7237                87   
19   12144                65   

                                                 text  
0   Admission Date:  [**2101-10-20**]     Discharg...  
1   Admission Date:  [**2191-3-16**]     Discharge...  
2   Admission Date: [**2175-5-30**]        Dischar...  
3   Name:  [**Known lastname 10050**], [**Known fi...  
4   Admission Date:  [**2149-11-9**]       Dischar...  
5   Admission Date:  [**2178-4-16**]              ...  
6   Admission Date:  [**2104-8-7**]     Discharge ...  
7   Name:  [**Known lastname 9900**], [**Known fir...  
8   Admission Date:  [**2167-1-8**]       Discharg...  
9   Admission Date:  [**2135-5-9**]              D...  
10  Admission Date:  [**2134-12-27**]             ...  
11  Admission Date:  [**2167-10-2**]              ...  
12  Admission Date:  [**2167-9-30**]              ...  
13  Name:  [**Known lastname 4919**] JR,[**Known f...  
14  Admission Date:  [**2108-8-5**]              D...  
15  Admission Date:  [**2183-4-28**]       Dischar...  
16  Admission Date:  [**2183-5-10**]       Dischar...  
17  Admission Date:  [**2135-1-30**]              ...  
18  Admission Date:  [**2134-9-11**]              ...  
19  Admission Date:  [**2196-4-9**]       Discharg...

In [5]:
df.shape

(55404, 9)

55404 'adult' (15 or over) discharge summaries - this is what we expect from our exploratory data analysis

The following punctuation marks frequently appear in the middle of words or between words without spacing meaning they are missed by the tokenizer. What we need to is to split the tokens on these punctuation marks after we have tokenized. We do this with regex. We then retokenize. This will substantially decreases the number of our unique words which we will replace with <UNK>

- ampersand
- brackets
- colons
- forward slashes(make sure to leave dates alone though)
- full stops
- hyphens
- equals signs
- semicolons
- plus signs

In [6]:
date_regex = re.compile(r'([0-9])-([0-9][0-9]?)-([0-9])') # change date format so spacy can recognise
newline_regex = re.compile(r'(\\n){3,}') # cap number of consecutive newline characters to 2
newline_regex2 = re.compile(r'(\\r){3,}') # cap number of consecutive newline characters to 2
ellipsis_regex = re.compile(r'(\.){2,}')
tilda_mult_regex = re.compile(r'(~){2,}')
atsign_mult_regex = re.compile(r'(@){2,}')

bracket_regex = re.compile(r'(.)(\()(.)')
bracket_regex2 = re.compile(r'(.)(\))(.)')
slash_regex = re.compile(r'(.)(\/)([^0-9])')
slash_regex2 = re.compile(r'([^0-9])(\/)(.)')
equals_regex = re.compile(r'(.)(=)(.)')
colon_regex = re.compile(r'(.)(:)(.)')
sq_bracket_regex = re.compile(r'(.)(\[)(.)')
dash_regex = re.compile(r'(.)(-)(.)')
dash_regex2 = re.compile(r'(-)([\S])')
plus_regex = re.compile(r'(.)(\+)(.)')
amp_regex = re.compile(r'(.)(&)(.)')
star_regex = re.compile(r'(.)(\*)(.)') 
comma_regex = re.compile(r'(.)(,)(.)')
tilda_regex = re.compile(r'(.)(~)(.)')
pipe_regex = re.compile(r'(.)(\|)(.)')
atsign_regex = re.compile(r'(.)(@)(.)')
dot_regex = re.compile(r'([^.][^0-9])(\.)([^0-9,][^.])')

dot_regex2 = re.compile(r'([^0-9])(\.)(.)')
semicol_regex = re.compile(r'(.);(.)')
caret_regex = re.compile(r'(.)\^(.)')

In [7]:
nlp = spacy.load('en_core_sci_md') # sciSpaCy

nlp.tokenizer.add_special_case(u'<PAR>', [{ORTH: u'<PAR>'}])
nlp.tokenizer.add_special_case(u'<UNK>', [{ORTH: u'<UNK>'}])

num_tokens = list()
num_sents = list()

i = 0

def tokenise_text(text, counter):
    global i
    
    text = str(text).lower()
    text = date_regex.sub(r'\1/\2/\3',text)
    text = newline_regex.sub(r' \\n\\n ',text)
    text = newline_regex2.sub(r' \\n\\n ',text)
    text = ellipsis_regex.sub(r'.',text)
    text = tilda_mult_regex.sub(r'~',text)
    text = atsign_mult_regex.sub(r'@',text)
    
    text = text.replace("[**","[").replace("**]","]")
    
    tokens = nlp.tokenizer(text)
    tokenised_text = ""
    
    for token in tokens:
        tokenised_text = tokenised_text + token.text + " "
    
    tokenised_text = tokenised_text.replace("\n"," <PAR> ")
    
    tokenised_text = bracket_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = bracket_regex2.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = slash_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = slash_regex2.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = slash_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = equals_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = colon_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = sq_bracket_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = dash_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = dash_regex.sub(r'\1 \2 \3',tokenised_text) # dash twice because sometimes it appears twice
    tokenised_text = dash_regex.sub(r'\1 \2 \3',tokenised_text) # dash thrice because sometimes it appears thrice
    tokenised_text = dash_regex2.sub(r'\1 \2',tokenised_text) # dash thrice because sometimes it appears thrice
    tokenised_text = plus_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = star_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = amp_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = comma_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = dot_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = atsign_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = tilda_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = pipe_regex.sub(r'\1 \2 \3',tokenised_text)
    tokenised_text = dot_regex2.sub(r'\1 \3',tokenised_text)
    tokenised_text = semicol_regex.sub(r'\1 \2',tokenised_text)
    tokenised_text = caret_regex.sub(r'\1 \2',tokenised_text)
        
    tokenised_text = ' '.join(tokenised_text.split())
    
    tokens = nlp.tokenizer(tokenised_text)
    tokenised_text = ""
    
    for token in tokens:
        tokenised_text = tokenised_text + token.text + " "
    
    counter.update(tokenised_text.split())
    tokenized_text = nltk.sent_tokenize(text)

    num_tokens.append(len(list(tokens)))
    num_sents.append(len(tokenized_text))
    
    i += 1
    if (i % 100) == 0:
        print (i)
    
    return tokenised_text

In [8]:
word_freq = Counter()
df["text"] = df["text"].apply(tokenise_text, args = (word_freq,))

100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
5300
5400
5500
5600
5700
5800
5900
6000
6100
6200
6300
6400
6500
6600
6700
6800
6900
7000
7100
7200
7300
7400
7500
7600
7700
7800
7900
8000
8100
8200
8300
8400
8500
8600
8700
8800
8900
9000
9100
9200
9300
9400
9500
9600
9700
9800
9900
10000
10100
10200
10300
10400
10500
10600
10700
10800
10900
11000
11100
11200
11300
11400
11500
11600
11700
11800
11900
12000
12100
12200
12300
12400
12500
12600
12700
12800
12900
13000
13100
13200
13300
13400
13500
13600
13700
13800
13900
14000
14100
14200
14300
14400
14500
14600
14700
14800
14900
15000
15100
15200
15300
15400
15500
15600
15700
15800
15900
16000
16100
16200
16300
16400
16500
16600
16700
16800
16900
17000
17100
17200
17300
17400
17500
17600
17700
17800
17900
18000
18100
18200
18300
18400
1850

Below we isolate the tokens which appear 3 times or fewer. They are mostly misspellings.

In [9]:
infreq_words = [word for word in word_freq.keys() if word_freq[word] <= 3 and word[0].isdigit() == False]
print(len(infreq_words))
sorted(infreq_words)[10000:10050]

95675


['blacktop',
 'blacouts',
 'bladdder',
 'bladde',
 'bladded',
 'bladderincontinence',
 'bladders',
 'bladderscan',
 'blah',
 'blaiteral',
 'blakemoor',
 'blakmore',
 'blanace',
 'blance',
 'blanch',
 'blancheable',
 'blanchible',
 'blancket',
 'blankes',
 'blanketing',
 'blanking',
 'blaoon',
 'blaoting',
 'blare',
 'blase',
 'blask',
 'blastemia',
 'blasting',
 'blastocystisi',
 'blastocystitis',
 'blastocytes',
 'blastoid',
 'blastoma',
 'blastomycoes',
 'blat',
 'blatant',
 'blatent',
 'blaterl',
 'blation',
 'blayton',
 'blbocavernous',
 'blc2',
 'blcoakde',
 'blcok',
 'blcokade',
 'blcoker',
 'bldder',
 'bldk',
 'bldlg',
 'bldneg']

We try and see if we can correct the misspellings using the `pyspellchecker` library by using the Levenshtein Distance algorithm and comparing against a dictionary. We first add the words with >3 occurrence to our dictionary. This is because they include a lot of scientific/medical terms which might not already be there

In [10]:
freq_words = [word for word in word_freq.keys() if word_freq[word] > 3]
add_to_dictionary = " ".join(freq_words)
f=open("../data/mimic_dict.txt", "w+")
f.write(add_to_dictionary)
f.close()

In [11]:
spell = SpellChecker()
spell.distance = 1  # set the distance parameter to just 1 edit away - much quicker
spell.word_frequency.load_text_file('../data/mimic_dict.txt')

In [12]:
misspelled = spell.unknown(infreq_words)
misspell_dict = {}
for i, word in enumerate(misspelled):
    if (word != spell.correction(word)):
        misspell_dict[word] = spell.correction(word)
    if (i % 100 == 0):
        print(i)

0
100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
5300
5400
5500
5600
5700
5800
5900
6000
6100
6200
6300
6400
6500
6600
6700
6800
6900
7000
7100
7200
7300
7400
7500
7600
7700
7800
7900
8000
8100
8200
8300
8400
8500
8600
8700
8800
8900
9000
9100
9200
9300
9400
9500
9600
9700
9800
9900
10000
10100
10200
10300
10400
10500
10600
10700
10800
10900
11000
11100
11200
11300
11400
11500
11600
11700
11800
11900
12000
12100
12200
12300
12400
12500
12600
12700
12800
12900
13000
13100
13200
13300
13400
13500
13600
13700
13800
13900
14000
14100
14200
14300
14400
14500
14600
14700
14800
14900
15000
15100
15200
15300
15400
15500
15600
15700
15800
15900
16000
16100
16200
16300
16400
16500
16600
16700
16800
16900
17000
17100
17200
17300
17400
17500
17600
17700
17800
17900
18000
18100
18200
18300
18400
18

In [13]:
print(len(misspell_dict))
list(misspell_dict.items())[:10]

61917


[('phenlyephrine', 'phenylephrine'),
 ('submanibular', 'submandibular'),
 ('vetral', 'ventral'),
 ('encepahlopathic', 'encephalopathic'),
 ('nuerologically', 'neurologically'),
 ('acress', 'across'),
 ('spepe', 'pepe'),
 ('arthrocenteis', 'arthrocentesis'),
 ('hydrochlorotiazide', 'hydrochlorothiazide'),
 ('witgh', 'with')]

We now have correct spellings for many words in our dictionary that occurred <= 3 times. Anything else will be marked as `UNK`. We will save these as text files to avoid having to run this computation again.

In [14]:
unk_words = [word for word in infreq_words if word not in list(misspell_dict.keys())]
print(len(unk_words))
unk_words[:100]

33758


['npg',
 'cytosis',
 'movlapine',
 'decongest',
 'neuronavigation',
 'nighthawk',
 'ego',
 'driverlives',
 'flegmon',
 'charcoaled',
 'facile',
 'oversensing',
 'arboviruses',
 'penotarbital',
 'uvulopalatotectomy',
 'dischargard',
 'resternotomy',
 'provatella',
 'n0.was',
 'exhaustible',
 'hemodymanic',
 'bronchcospy',
 'dermotalogy',
 'd10/14',
 'd10/42',
 'd5/14',
 'skidding',
 'fena>1',
 'wayside',
 'x4mos',
 'percocetq4',
 'arthridities',
 'soonafter',
 'silpiride',
 'deoxyribonucleotide',
 'jetted',
 'remeasurement',
 'crave',
 'stuporousness',
 'targretin',
 'pyracin',
 'thrombocytonia',
 'capitol',
 'xylose',
 'nonulcerative',
 'polyradiculopathies',
 'r12/40',
 'uop>30cc',
 'decapsulation',
 'visitations',
 'francetta',
 'embolect',
 'bilayer',
 'calcinet',
 'satuth',
 'porphyrobilinogen',
 'grouping',
 'addendings',
 'weregiven',
 'nifepidine',
 'drafted',
 'titled',
 'bp253/170',
 'supratherap',
 'bp260/130',
 'bp176/135',
 'microthrombosis',
 'admissionwho',
 'coolata',
 '

In [15]:
if 'treatement' in misspell_dict.keys():
    print (misspell_dict['treatement'])

In [16]:
np.savetxt('../data/discharge_unk_words.txt', unk_words, fmt='%s', newline=os.linesep)

f=open("../data/discharge_typos.txt", "w+")
for key in misspell_dict:
    f.write(key + '\t' + misspell_dict[key] + '\n')
f.close()

In [17]:
# LOAD TYPOS AND UNKS FROM TXT FILE IF ALREADY GENERATED

#discharge_typos = pd.read_csv('../data/disharge_typos.txt', sep='\t',header=None)
#misspell_dict = {row[0]: row[1] for row in discharge_typos.values}
#unk_words = pd.read_csv('../data/discharge_unk_words.txt', sep='\t', header=None)
#unk_words = list(unk_words[0])

 We will correct spelling mistakes whilst any word left uncorrected will be replaced with `<UNK>`. Instead of using regex, we instead use `FlashText` which is considerably faster for replacing words from a large dictionary.

In [18]:
keywords = list(misspell_dict.keys())
clean = list(misspell_dict.values())

processor = KeywordProcessor()

for keyword_name, clean_name in zip(keywords, clean):
    processor.add_keyword(keyword_name, clean_name)
    
for unk in unk_words:
    processor.add_keyword(unk, "<UNK>")

counter = 0
def fix_typos(text):
    global counter
    
    found = processor.replace_keywords(text)
    
    counter+=1
    if (counter % 100) == 0:
        print (counter)
    
    return found


# apply tokenising function elementwise
df["text"] = df["text"].apply(fix_typos)
df.head()

100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
5300
5400
5500
5600
5700
5800
5900
6000
6100
6200
6300
6400
6500
6600
6700
6800
6900
7000
7100
7200
7300
7400
7500
7600
7700
7800
7900
8000
8100
8200
8300
8400
8500
8600
8700
8800
8900
9000
9100
9200
9300
9400
9500
9600
9700
9800
9900
10000
10100
10200
10300
10400
10500
10600
10700
10800
10900
11000
11100
11200
11300
11400
11500
11600
11700
11800
11900
12000
12100
12200
12300
12400
12500
12600
12700
12800
12900
13000
13100
13200
13300
13400
13500
13600
13700
13800
13900
14000
14100
14200
14300
14400
14500
14600
14700
14800
14900
15000
15100
15200
15300
15400
15500
15600
15700
15800
15900
16000
16100
16200
16300
16400
16500
16600
16700
16800
16900
17000
17100
17200
17300
17400
17500
17600
17700
17800
17900
18000
18100
18200
18300
18400
1850

subject_id        dob gender  hadm_id           category  chartdate  \
0           3 2025-04-11      M   145834  Discharge summary 2101-10-31   
1           4 2143-05-12      F   185777  Discharge summary 2191-03-23   
2           6 2109-06-21      F   107064  Discharge summary 2175-06-15   
3           9 2108-01-26      M   150750  Discharge summary 2149-11-14   
4           9 2108-01-26      M   150750  Discharge summary 2149-11-13   

   row_id  age_at_noteevent                                               text  
0   44005                77  admission date : [ 2101/10/20 ] discharge date...  
1    4788                48  admission date : [ 2191/3/16 ] discharge date ...  
2   20825                66  admission date : [ 2175/5/30 ] discharge date ...  
3   57115                42  name : [ known lastname 10050 ] , [ known firs...  
4   20070                42  admission date : [ 2149/11/9 ] discharge date ...

In [19]:
' '.join(df.head(20)['text'])

'admission date : [ 2101/10/20 ] discharge date : [ 2101/10/31 ] <PAR> <PAR> date of birth : [ 2025/4/11 ] sex : m <PAR> <PAR> service : medicine <PAR> <PAR> chief complaint : admitted from rehabilitation for <PAR> hypotension ( systolic blood pressure to the 70s ) and <PAR> decreased urine output <PAR> <PAR> history of present illness : the patient is a 76 - year - old <PAR> male who had been hospitalized at the [ hospital1 190 ] from [ 10 - 11 ] through [ 10 - 19 ] of [ 2101 ] <PAR> after undergoing a left femoral - at bypass graft and was <PAR> subsequently discharged to a rehabilitation facility <PAR> <PAR> on [ 2101/10/20 ] , he presented again to the [ hospital1 346 ] after being found to have a systolic <PAR> blood pressure in the 70s and no urine output for 17 hours <PAR> a foley catheter placed at the rehabilitation facility <PAR> yielded 100 cc of murky / brown urine there may also have <PAR> been purulent discharge at the penile meatus at this time <PAR> <PAR> on presentatio

The text field has now been fully cleaned and tokenised. We can proceed to extract the first few tokens to use as a hint and move forth with joining other tables and partitioning the dataset.

In [20]:
counter = 0
def produce_hint(text):
    global counter
    l = text.split()
    counter += 1
    if (counter % 10000) == 0:
        print (counter)
    return ' '.join(l[:10]) # first 10 tokens

df['hint'] = df['text'].map(lambda x: produce_hint(x))

print(df.shape)
df.head()

10000
20000
30000
40000
50000
(55404, 10)


subject_id        dob gender  hadm_id           category  chartdate  \
0           3 2025-04-11      M   145834  Discharge summary 2101-10-31   
1           4 2143-05-12      F   185777  Discharge summary 2191-03-23   
2           6 2109-06-21      F   107064  Discharge summary 2175-06-15   
3           9 2108-01-26      M   150750  Discharge summary 2149-11-14   
4           9 2108-01-26      M   150750  Discharge summary 2149-11-13   

   row_id  age_at_noteevent  \
0   44005                77   
1    4788                48   
2   20825                66   
3   57115                42   
4   20070                42   

                                                text  \
0  admission date : [ 2101/10/20 ] discharge date...   
1  admission date : [ 2191/3/16 ] discharge date ...   
2  admission date : [ 2175/5/30 ] discharge date ...   
3  name : [ known lastname 10050 ] , [ known firs...   
4  admission date : [ 2149/11/9 ] discharge date ...   

                                                hint  
0  admission date : [ 2101/10/20 ] discharge date...  
1  admission date : [ 2191/3/16 ] discharge date : [  
2  admission date : [ 2175/5/30 ] discharge date : [  
3          name : [ known lastname 10050 ] , [ known  
4  admission date : [ 2149/11/9 ] discharge date : [

In [21]:
# patients above 89 years of age had their dob modified to be 300 years old at time of first event for privacy reasons
# change their age to instead be 90

df.loc[df['age_at_noteevent'] > 200, 'age_at_noteevent'] = 90

We now merge our dataframe with some extra information from the admissions table. This preprocessing has already been done in another notebook so here we just load the csv file. Primarily, this table provides us with the ethnicity of the patient and whether or not they go on to have an unplanned readmission within 30 days of being discharged. This is relevant for one of our downstream tasks using the artificial data

In [4]:
# admissions data

admissions = pd.read_csv('../data/df_adm.csv', sep=',')
admissions.head()
df = pd.merge(df, admissions,  how='left', left_on=['subject_id','hadm_id'], right_on = ['subject_id','hadm_id'])

Now we load all the context data from other tables. We'll be using this to construct the structured input to our encoder

In [23]:
# lab items data

df_labitems = pd.read_sql_query('''
  SELECT l.subject_id, l.charttime, l.value, l.valueuom, l.flag, d.label
  FROM labevents l
  INNER JOIN d_labitems d 
  USING (itemid)
  ORDER BY l.subject_id
  --LIMIT 10000;
''', cnx)

print(df_labitems.shape)
df_labitems.head()

(27854055, 6)


subject_id           charttime value valueuom      flag  \
0           2 2138-07-17 20:48:00     0        %      None   
1           2 2138-07-17 20:48:00     0        %      None   
2           2 2138-07-17 20:48:00     0        %      None   
3           2 2138-07-17 20:48:00     0        %      None   
4           2 2138-07-17 20:48:00     0        %  abnormal   

                  label  
0  Atypical Lymphocytes  
1                 Bands  
2             Basophils  
3           Eosinophils  
4            Hematocrit

In [24]:
# prescriptions data

df_prescriptions = pd.read_sql_query('''
  SELECT subject_id, startdate, enddate, LOWER(drug), prod_strength
  FROM prescriptions
  ORDER BY subject_id
  --LIMIT 10000;
''', cnx)

print(df_prescriptions.shape)
df_prescriptions.head()

(4156450, 5)


subject_id  startdate    enddate                          lower  \
0           2 2138-07-18 2138-07-20              neo*iv*gentamicin   
1           2 2138-07-18 2138-07-20       syringe (neonatal) *d5w*   
2           2 2138-07-18 2138-07-21              ampicillin sodium   
3           2 2138-07-18 2138-07-21                send 500mg vial   
4           4 2191-03-16 2191-03-23  guaifenesin-codeine phosphate   

     prod_strength  
0      10mg/mL-2mL  
1        1 Syringe  
2       500mg Vial  
3  Send 500mg Vial  
4        5ML UDCUP

In [25]:
# diagnoses data

df_diagnoses = pd.read_sql_query('''
  SELECT d.subject_id, d.hadm_id, d.seq_num, d.icd9_code, icd.short_title, LOWER(icd.long_title)
  FROM diagnoses_icd d
  INNER JOIN d_icd_diagnoses icd 
  USING (icd9_code)
  ORDER BY d.subject_id, d.seq_num
  --LIMIT 10000;
''', cnx)

print(df_diagnoses.shape)
df_diagnoses.head(10)

(634709, 6)


subject_id  hadm_id  seq_num icd9_code               short_title  \
0           2   163353        1     V3001    Single lb in-hosp w cs   
1           2   163353        2      V053  Need prphyl vc vrl hepat   
2           2   163353        3      V290    NB obsrv suspct infect   
3           3   145834        1      0389            Septicemia NOS   
4           3   145834        2     78559      Shock w/o trauma NEC   
5           3   145834        3      5849  Acute kidney failure NOS   
6           3   145834        4      4275            Cardiac arrest   
7           3   145834        5     41071  Subendo infarct, initial   
8           3   145834        6      4280                   CHF NOS   
9           3   145834        7      6826         Cellulitis of leg   

                                               lower  
0  single liveborn, born in hospital, delivered b...  
1  need for prophylactic vaccination and inoculat...  
2     observation for suspected infectious condition  
3                             unspecified septicemia  
4              other shock without mention of trauma  
5                  acute kidney failure, unspecified  
6                                     cardiac arrest  
7  subendocardial infarction, initial episode of ...  
8              congestive heart failure, unspecified  
9         cellulitis and abscess of leg, except foot

In [26]:
# procedures data

hadm_proc_subset = list(pd.read_csv('../data/df_proc_hadm_ids.csv', sep=',')['hadm_id'])

df_procedures = pd.read_sql_query('''
  SELECT p.subject_id, p.hadm_id, p.seq_num, p.icd9_code, icd.short_title, LOWER(icd.long_title)
  FROM procedures_icd p
  INNER JOIN d_icd_procedures icd 
  USING (icd9_code)
  ORDER BY p.subject_id, p.seq_num DESC
  --LIMIT 10000;
''', cnx)

print(df_procedures.shape)
df_procedures = df_procedures[df_procedures['hadm_id'].isin(hadm_proc_subset)]
print(df_procedures.shape)
df_procedures.head(20)

(237948, 6)
(212258, 6)


subject_id  hadm_id  seq_num icd9_code               short_title  \
1            3   145834        6       966   Entral infus nutrit sub   
2            3   145834        5      3893           Venous cath NEC   
3            3   145834        4      9672  Cont inv mec ven 96+ hrs   
4            3   145834        3      8964  Pulmon art wedge monitor   
5            3   145834        2      9962    Heart countershock NEC   
6            3   145834        1      9604  Insert endotracheal tube   
7            4   185777        3      3323        Other bronchoscopy   
8            4   185777        2      8872       Dx ultrasound-heart   
9            4   185777        1      3893           Venous cath NEC   
11           6   107064        5      9904   Packed cell transfusion   
12           6   107064        4      3806   Abdomen artery incision   
13           6   107064        3      3957    Rep vess w synth patch   
14           6   107064        2      0091   Trnsplnt live rel donor   
15           6   107064        1      5569     Kidney transplant NEC   
20           9   150750        2      9604  Insert endotracheal tube   
21           9   150750        1      9672  Cont inv mec ven 96+ hrs   
25          11   194540        5      9925   Inject ca chemother NEC   
26          11   194540        4      9229   Radiotherapeut proc NEC   
27          11   194540        3      0113       Closed brain biopsy   
28          11   194540        2      0113       Closed brain biopsy   

                                                lower  
1   enteral infusion of concentrated nutritional s...  
2    venous catheterization, not elsewhere classified  
3   continuous invasive mechanical ventilation for...  
4                   pulmonary artery wedge monitoring  
5                other electric countershock of heart  
6                      insertion of endotracheal tube  
7                                  other bronchoscopy  
8                      diagnostic ultrasound of heart  
9    venous catheterization, not elsewhere classified  
11                        transfusion of packed cells  
12             incision of vessel, abdominal arteries  
13  repair of blood vessel with synthetic patch graft  
14                 transplant from live related donor  
15                       other kidney transplantation  
20                     insertion of endotracheal tube  
21  continuous invasive mechanical ventilation for...  
25  injection or infusion of cancer chemotherapeut...  
26                   other radiotherapeutic procedure  
27     closed [percutaneous] [needle] biopsy of brain  
28     closed [percutaneous] [needle] biopsy of brain

In [27]:
# microbiology data

df_microbiology = pd.read_sql_query('''
  SELECT 
      subject_id, 
      hadm_id, 
      MAX(chartdate) AS chartdate, 
      LOWER(spec_type_desc), 
      string_agg(DISTINCT(LOWER(org_name)), ', ') AS organism
  
  FROM microbiologyevents
  GROUP BY subject_id, hadm_id, charttime, spec_type_desc
  ORDER BY subject_id, chartdate DESC
  --LIMIT 10000;
''', cnx)

print(df_microbiology.shape)
df_microbiology.head(20)

(340305, 5)


subject_id  hadm_id  chartdate                                 lower  \
0            2   163353 2138-07-17               blood culture - neonate   
1            3   145834 2101-10-28                                 stool   
2            3   145834 2101-10-26                                 urine   
3            3   145834 2101-10-21                         blood culture   
4            3   145834 2101-10-21                                 urine   
5            3   145834 2101-10-21                       catheter tip-iv   
6            3   145834 2101-10-21                         blood culture   
7            3   145834 2101-10-21  blood culture ( myco/f lytic bottle)   
8            3   145834 2101-10-21                         blood culture   
9            3   145834 2101-10-21                                sputum   
10           3   145834 2101-10-20                         blood culture   
11           3   145834 2101-10-20                                  swab   
12           3   145834 2101-10-20                         blood culture   
13           4   185777 2191-03-23                                 stool   
14           4   185777 2191-03-22                         blood culture   
15           4   185777 2191-03-22                         blood culture   
16           4   185777 2191-03-19                                sputum   
17           4   185777 2191-03-18                         blood culture   
18           4   185777 2191-03-18                         blood culture   
19           4   185777 2191-03-18                         blood culture   

                                             organism  
0                                                None  
1                                                None  
2                                               yeast  
3                                                None  
4                                               yeast  
5                                                None  
6                                                None  
7                                                None  
8                                                None  
9                                               yeast  
10                                               None  
11  gram negative rod(s), probable enterococcus, s...  
12                                               None  
13                                               None  
14                                               None  
15                                               None  
16                                               None  
17                                               None  
18                                               None  
19                                               None

In [28]:
# phenotype classification subject - we need to make sure these are part of our test set

annotations = pd.read_csv('../data/annotations.csv')
annotations.columns.values[0] = 'hadm_id'
annotations.columns.values[1] = 'subject_id'

pheno_subjects = list(annotations['subject_id'])
print(annotations.shape)
annotations.head()

(1610, 18)


hadm_id  subject_id  chart.time  cohort  Obesity  Non.Adherence  \
0   118003        3644      118003       1        0              0   
1   177830       97736      999999       0        0              0   
2   185673       27694      999999       0        0              0   
3   131938       16275      131938       0        0              0   
4   198999        4059      198999       1        0              0   

   Developmental.Delay.Retardation  Advanced.Heart.Disease  \
0                                0                       0   
1                                0                       0   
2                                0                       0   
3                                0                       0   
4                                0                       0   

   Advanced.Lung.Disease  Schizophrenia.and.other.Psychiatric.Disorders  \
0                      0                                              0   
1                      0                                              0   
2                      0                                              0   
3                      0                                              0   
4                      0                                              1   

   Alcohol.Abuse  Other.Substance.Abuse  Chronic.Pain.Fibromyalgia  \
0              0                      0                          1   
1              0                      0                          0   
2              0                      0                          0   
3              0                      0                          0   
4              0                      0                          0   

   Chronic.Neurological.Dystrophies  Advanced.Cancer  Depression  Dementia  \
0                                 0                0           1         0   
1                                 0                1           0         0   
2                                 0                0           1         0   
3                                 0                0           0         0   
4                                 0                0           1         0   

   Unsure  
0       0  
1       0  
2       0  
3       0  
4       0

We now split the dataset into training, validation and test sets. As we do so, we query the relevant context data and append it to the input

In [29]:
#%%timeit -n 3 -r 3

# Split the dataset in a grouped and stratified manner

def StratifiedGroupShuffleSplit(df_main):
    
    np.random.seed(768) # seeded for reproducibility
    df_main = df_main.reindex(np.random.permutation(df_main.index)) # shuffle dataset
    
    # create empty train, val and test datasets
    df_train = pd.DataFrame()
    df_val = pd.DataFrame()
    df_test = pd.DataFrame()

    hparam_mse_wgt = 0.8 # must be between 0 and 1
    assert(0 <= hparam_mse_wgt <= 1)
    train_proportion = 0.8 # must be between 0 and 1
    assert(0 <= train_proportion <= 1)
    val_test_proportion = (1-train_proportion)/2

    subject_grouped_df_main = df_main.groupby(['subject_id'], sort=False, as_index=False)
    readmit_grouped_df_main = df_main.groupby('30d_unplan_readmit').count()[['subject_id']]/len(df_main)*100 
    
    # function to calculate loss
    def calc_mse_loss(df):
        grouped_df = df.groupby('30d_unplan_readmit').count()[['subject_id']]/len(df)*100
        df_temp = readmit_grouped_df_main.join(grouped_df, on = '30d_unplan_readmit', how = 'left', lsuffix = '_main')
        df_temp.fillna(0, inplace=True)
        df_temp['diff'] = (df_temp['subject_id_main'] - df_temp['subject_id'])**2
        mse_loss = np.mean(df_temp['diff'])
        return mse_loss
    
    directory = "../data/preprocessed/"
    
    f_train = open(directory + "src-train.txt","w+")
    f_val = open(directory + "src-val.txt","w+")
    f_test = open(directory + "src-test.txt","w+")
    
    len_train = 0
    len_val = 0
    len_test = 0
    total_records = 0
    i = 0

    # loop the groups of subjects one by one
    for _, group in subject_grouped_df_main:

        total_records = len_train + len_val + len_test
        g = pd.DataFrame(group)
        subject_id = g['subject_id'].iloc[0]
        
        pre_left = df_prescriptions['subject_id'].searchsorted(subject_id, 'left')
        pre_right = df_prescriptions['subject_id'].searchsorted(subject_id, 'right')
        
        lab_left = df_labitems['subject_id'].searchsorted(subject_id, 'left')
        lab_right = df_labitems['subject_id'].searchsorted(subject_id, 'right')
        
        diag_left = df_diagnoses['subject_id'].searchsorted(subject_id, 'left')
        diag_right = df_diagnoses['subject_id'].searchsorted(subject_id, 'right')
        
        proc_left = df_procedures['subject_id'].searchsorted(subject_id, 'left')
        proc_right = df_procedures['subject_id'].searchsorted(subject_id, 'right')
        
        micro_left = df_microbiology['subject_id'].searchsorted(subject_id, 'left')
        micro_right = df_microbiology['subject_id'].searchsorted(subject_id, 'right')
        
        g_prescriptions = df_prescriptions[pre_left:pre_right]
        g_labitems = df_labitems[lab_left:lab_right]
        g_diagnoses = df_diagnoses[diag_left:diag_right]
        g_procedures = df_procedures[proc_left:proc_right]
        g_microbiology = df_microbiology[micro_left:micro_right]
        
        train = False
        val = False
        test = False
        i += 1
        
        # all subjects in the phenotyping dataset need to be in the test set
        if subject_id in pheno_subjects:
            df_test = df_test.append(g, ignore_index=True)
            len_test += len(g)
            test = True
            
        # just to add something to each group to start off with - otherwise we end up dividing by 0
        elif (len_train == 0 or len_val == 0 or len_test == 0):
            if (len_train == 0):
                df_train = df_train.append(g, ignore_index=True)
                len_train += len(g)
                train = True
            elif (len_val == 0):
                df_val = df_val.append(g, ignore_index=True)
                len_val += len(g)
                val = True
            else:
                df_test = df_test.append(g, ignore_index=True)
                len_test += len(g)
                test = True
        
        # every 3rd group, balance the groups jointly by proportion and by ratio of unplanned readmissions
        elif (i % 3 == 0):
            
            mse_loss_diff_train = calc_mse_loss(df_train) - calc_mse_loss(df_train.append(g, ignore_index=True))
            mse_loss_diff_val = calc_mse_loss(df_val) - calc_mse_loss(df_val.append(g, ignore_index=True))
            mse_loss_diff_test = calc_mse_loss(df_test) - calc_mse_loss(df_test.append(g, ignore_index=True))

            len_diff_train = (train_proportion - (len_train/total_records)) * 100
            len_diff_val = (val_test_proportion - (len_val/total_records)) * 100
            len_diff_test = (val_test_proportion - (len_test/total_records)) * 100

            len_loss_diff_train = len_diff_train * abs(len_diff_train)
            len_loss_diff_val = len_diff_val * abs(len_diff_val)
            len_loss_diff_test = len_diff_test * abs(len_diff_test)

            loss_train = (hparam_mse_wgt * mse_loss_diff_train) + ((1-hparam_mse_wgt) * len_loss_diff_train)
            loss_val = (hparam_mse_wgt * mse_loss_diff_val) + ((1-hparam_mse_wgt) * len_loss_diff_val)
            loss_test = (hparam_mse_wgt * mse_loss_diff_test) + ((1-hparam_mse_wgt) * len_loss_diff_test)

            if (max(loss_train,loss_val,loss_test) == loss_train):
                df_train = df_train.append(g, ignore_index=True)
                len_train += len(g)
                train = True
            elif (max(loss_train,loss_val,loss_test) == loss_val):
                df_val = df_val.append(g, ignore_index=True)
                len_val += len(g)
                val = True
            else:
                df_test = df_test.append(g, ignore_index=True)
                len_test += len(g)
                test = True
            
            if i % 100 == 0:
                print ("Group " + str(i) + ": loss_train: " + str(loss_train) + " | " + 
                                             "loss_val: " + str(loss_val) + " | " + 
                                             "loss_test: " + str(loss_test))

        # all the other groups - divided simply by ratios of the dataset splits
        else:
            train_diff = train_proportion - (len_train/total_records)
            val_diff = val_test_proportion - (len_val/total_records)
            test_diff = val_test_proportion - (len_test/total_records)
            
            if (max(train_diff, val_diff, test_diff) == train_diff):
                df_train = df_train.append(g, ignore_index=True)
                len_train += len(g)
                train = True
            elif (max(train_diff, val_diff, test_diff) == val_diff):
                df_val = df_val.append(g, ignore_index=True)
                len_val += len(g)
                val = True
            else:
                df_test = df_test.append(g, ignore_index=True)
                len_test += len(g)
                test = True
        
        # loop through every row in the group to get relevant prescriptions and lab items before appending to file
        for j, row in enumerate(g.itertuples()):
            
            hadm_id = row[4]
            # context is all day the day before the discharge day and the day of the discharge
            chartdate = datetime.combine(row[6], datetime.min.time()) + timedelta(days=1)
            cutoff = chartdate - timedelta(days=2)
            micro_cutoff = chartdate - timedelta(days=4) # greater context window for microbiology results
            
            ###### LAB RESULTS
            lab_condition = np.logical_and((g_labitems.charttime >= cutoff),
                                           (g_labitems.charttime < chartdate))
            lab_items = g_labitems[lab_condition]
            lab_items = lab_items.sort_values(by=['charttime'], ascending=False)
            
            ###### PRESCRIPTIONS
            pre_condition = np.logical_and((g_prescriptions.startdate >= cutoff),
                                           (g_prescriptions.startdate < chartdate))
            prescriptions = g_prescriptions[pre_condition]
            prescriptions = prescriptions.sort_values(by=['startdate'], ascending=False)
            
            ###### DIAGNOSES
            diagnoses = g_diagnoses[g_diagnoses.hadm_id == hadm_id]
            diagnoses = diagnoses.sort_values(by=['seq_num'], ascending=True)
            
            ###### PROCEDURES
            if j == 0:
                procedures = g_procedures[g_procedures.hadm_id == hadm_id]
                procedures = procedures.sort_values(by=['seq_num'], ascending=False)
            
            ###### MICROBIOLOGY TESTS
            micro_condition = np.logical_and((g_microbiology.chartdate >= micro_cutoff),
                                             (g_microbiology.chartdate < chartdate))
            microbiology = g_microbiology[micro_condition]
            microbiology = microbiology.sort_values(by=['chartdate'], ascending=False)
            
            ##########################################
            lab_items_list = ""
            lab_items_length = len(lab_items)
            if (lab_items_length > 0):
                for k, lab_row in enumerate(lab_items.itertuples()):
                    flag = ""
                    if (pd.isna(lab_row[5]) == False):
                        flag = " , " + str(lab_row[5])

                    lab_items_list += str(lab_row[6]) + " , " + str(lab_row[3]) + " , " + str(lab_row[4]) + flag
                    if (k != (lab_items_length - 1)):
                        lab_items_list += " | "
            
            ##########################################
            prescriptions_list = ""
            prescriptions_length = len(prescriptions)
            if (prescriptions_length > 0):
                for k, pre_row in enumerate(prescriptions.itertuples()):
                    prescriptions_list += str(pre_row[4]) + " , " + str(pre_row[5])
                    if (k != (prescriptions_length - 1)):
                        prescriptions_list += " | "

            ##########################################
            diagnoses_list = ""
            diagnoses_length = len(diagnoses)
            if (diagnoses_length > 0):
                for k, diag_row in enumerate(diagnoses.itertuples()):
                    diagnoses_list += str(diag_row[6])
                    if (k != (diagnoses_length - 1)):
                        diagnoses_list += " | "

            ##########################################
            procedures_list = ""
            if j == 0: # only allow procedures for the most recent hospital admission for a given subject
                procedures_length = len(procedures)
                if (procedures_length > 0):
                    for k, proc_row in enumerate(procedures.itertuples()):
                        procedures_list += str(proc_row[6])
                        if (k != (procedures_length - 1)):
                            procedures_list += " | "

            ##########################################
            microbiology_list = ""
            microbiology_length = len(microbiology)
            if (microbiology_length > 0):
                for k, micro_row in enumerate(microbiology.itertuples()):
                    microbiology_list += str(micro_row[4]) + " : " + str(micro_row[5])
                    if (k != (microbiology_length - 1)):
                        microbiology_list += " | "
                        
            ##########################################
            
            if (train == True):
                f_train.write(str(row[10]) + " <H> " + str(row[3]) + " <G> " + str(row[8]) + " <A> " + 
                    str(row[11]) + " <E> " + diagnoses_list + " <D> " + procedures_list + " <P> " +
                    prescriptions_list + " <M> " + microbiology_list + " <T> " + lab_items_list + " <L>" + "\n")
            elif (val == True):
                f_val.write(str(row[10]) + " <H> " + str(row[3]) + " <G> " + str(row[8]) + " <A> " + 
                    str(row[11]) + " <E> " + diagnoses_list + " <D> " + procedures_list + " <P> " +
                    prescriptions_list + " <M> " + microbiology_list + " <T> " + lab_items_list + " <L>" + "\n")
            else:
                f_test.write(str(row[10]) + " <H> " + str(row[3]) + " <G> " + str(row[8]) + " <A> " + 
                    str(row[11]) + " <E> " + diagnoses_list + " <D> " + procedures_list + " <P> " +
                    prescriptions_list + " <M> " + microbiology_list + " <T> " + lab_items_list + " <L>" + "\n")

    f_train.close()
    f_val.close()
    f_test.close()
    
    return df_train, df_val, df_test

In [30]:
src_train, src_val, src_test = StratifiedGroupShuffleSplit(df)

Group 300: loss_train: 3.483879532489648 | loss_val: -5.065794711175795 | loss_test: -40.251142326009344
Group 600: loss_train: 3.485944564635249 | loss_val: 3.2425335954806545 | loss_test: -9.805147160843074
Group 900: loss_train: 3.9121727219329245 | loss_val: 3.023275795604748 | loss_test: -11.619612386492095
Group 1200: loss_train: 3.200839204847737 | loss_val: 3.228739822594459 | loss_test: -7.2532212788553005
Group 1800: loss_train: 1.7680622712216616 | loss_val: 1.131636523808363 | loss_test: -13.013579268879546
Group 2100: loss_train: 2.430805501192729 | loss_val: 2.218848541839631 | loss_test: -8.368964587623205
Group 2400: loss_train: 2.323659572741037 | loss_val: 2.2209325397333606 | loss_test: -7.480159185541306
Group 2700: loss_train: 1.826968534892006 | loss_val: 1.8205464380545573 | loss_test: -6.587310664502842
Group 3000: loss_train: 1.7786192783449686 | loss_val: 1.710734217780508 | loss_test: -6.293973636790505
Group 3300: loss_train: 1.6957191460695586 | loss_val: 1

Group 24900: loss_train: 0.027661635271785306 | loss_val: 0.02744113691079229 | loss_test: -0.041502670310753664
Group 25200: loss_train: 0.02556523439493033 | loss_val: 0.025351567519069303 | loss_test: -0.03461323286890872
Group 25500: loss_train: 0.02572798171031987 | loss_val: 0.025561491984204564 | loss_test: -0.037605170337595556
Group 25800: loss_train: 0.025068390325559582 | loss_val: 0.02452710789444393 | loss_test: -0.03541092995936475
Group 26100: loss_train: 0.025095817017317133 | loss_val: 0.023928268706194914 | loss_test: -0.036015528557097165
Group 26400: loss_train: 0.02363787740074864 | loss_val: 0.02208706623148853 | loss_test: 0.030378482912398946
Group 26700: loss_train: 0.02437178570428782 | loss_val: 0.023196302017590438 | loss_test: -0.03605418344924924
Group 27000: loss_train: 0.022850760111819423 | loss_val: 0.022652990716050897 | loss_test: -0.03326530272828335
Group 27300: loss_train: 0.02264732813437064 | loss_val: 0.022227915434219977 | loss_test: -0.033555

Now that we have finally split the dataset into training, evaluation and test datasets, let's just inspect the stratification to ensure that we have successfully managed to have a uniform ratio of unplanned readmissions across all three datasets. And that they have been split correctly in the ratio 8:1:1

In [44]:
print(len(src_train), len(src_val), len(src_test))

readmit_grouped_df_main = df.groupby('30d_unplan_readmit').count()[['subject_id']]/len(df)*100
grouped_train = src_train.groupby('30d_unplan_readmit').count()[['subject_id']]/len(src_train)*100
grouped_val = src_val.groupby('30d_unplan_readmit').count()[['subject_id']]/len(src_val)*100
grouped_test = src_test.groupby('30d_unplan_readmit').count()[['subject_id']]/len(src_test)*100

df_summary = readmit_grouped_df_main.join(grouped_train, on = '30d_unplan_readmit', how = 'left', rsuffix = '_train')
df_summary = df_summary.join(grouped_val, on = '30d_unplan_readmit', how = 'left', rsuffix = '_val')
df_summary = df_summary.join(grouped_test, on = '30d_unplan_readmit', how = 'left', rsuffix = '_test')
df_summary.fillna(0, inplace=True)
df_summary

44230 5447 5727


subject_id  subject_id_train  subject_id_val  \
30d_unplan_readmit                                                 
N                    93.839795         94.768257       94.125207   
Y                     6.160205          5.231743        5.874793   

                    subject_id_test  
30d_unplan_readmit                   
N                         86.397765  
Y                         13.602235

The ratio of 8:1:1 is roughly correct. The ratio of unplanned readmissions is also not too bad. However, the test dataset seems to contain considerably more unplanned readmissions: 13% vs 6%. 

In [52]:
print("# of Discharge summaries for subjects in phenotyping dataset: ", len(src_test[src_test['subject_id'].isin(pheno_subjects)]))
print("# of records in the phenotyping dataset",len(pheno_subjects))
print("# of unique subjects in the phenotyping dataset",len(annotations.subject_id.unique()))

# of Discharge summaries for subjects in phenotyping dataset:  3672
# of records in the phenotyping dataset 1610
# of unique subjects in the phenotyping dataset 1045


On closer inspection, this appears to be due to the fact that we require all subjects in the phenotyping dataset to be in the test set. Despite comprising only 1045 unique subjects, these subjects are responsible for 3672 discharge summaries - much higher than average. This naturally also corresponds to a much greater incidence of 30d unplanned readmission, hence the greater incidence for this in the test dataset. The difference between 13% and 6% is small enough that we will tolerate it.

We have saved our source datasets, now let's now save our target and reference datasets: 

In [53]:
# create target files

tgt_train = pd.DataFrame(src_train, columns = ["text"])
tgt_val = pd.DataFrame(src_val, columns = ["text"])
tgt_test = pd.DataFrame(src_test, columns = ["text"])

print(tgt_test.shape)
tgt_test.head()

(5727, 1)


text
0  admission date : [ 2164/8/31 ] discharge date ...
1  admission date : [ 2135/4/21 ] discharge date ...
2  admission date : [ 2138/10/16 ] discharge date...
3  admission date : [ 2133/11/27 ] discharge date...
4  admission date : [ 2139/7/22 ] discharge date ...

In [54]:
# save target files to disk

directory = "../data/preprocessed/"

np.savetxt(directory + 'tgt-train.txt', tgt_train, fmt='%s', newline=os.linesep)
np.savetxt(directory + 'tgt-val.txt', tgt_val, fmt='%s', newline=os.linesep)
np.savetxt(directory + 'tgt-test.txt', tgt_test, fmt='%s', newline=os.linesep)

In [61]:
# save reference files to disk

src_train.to_csv(directory + 'ref_train.tsv', sep='\t', index = False)
src_val.to_csv(directory + 'ref_val.tsv', sep='\t', index = False)
src_test.to_csv(directory + 'ref_test.tsv', sep='\t', index = False)